packages/otel/src/otel-exchange.ts

```diff
@@ -0,0 +1,74 @@
+import { type Span, SpanKind, SpanStatusCode, context } from "@opentelemetry/api";
+import { type CombinedError, type Operation, makeOperation, mapExchange } from "urql";
+import { getOtelTracer } from "./otel-tracer";
+import { AttributeNames, ObservabilityAttributes } from "./lib/observability-attributes";
+import { addInputVariableAttributes, addRequestHeaderAttributes } from "./otel-graphql-utils";
+import { SemanticAttributes } from "@opentelemetry/semantic-conventions";
+
+type Definition = {
+  name: {
+    value: string;
+  };
+};
+
+interface ExtendedFetchOptions extends RequestInit {
+  headers: Record<string, string>;
+}
+
+type ExtendedOperationContext = Operation["context"] & {
+  span: Span;
+  fetchOptions?: ExtendedFetchOptions;
+};
+
+interface ExtendedOperation extends Operation {
+  context: ExtendedOperationContext;
+}
+
+export const otelExchange = mapExchange({
+  onOperation(operation: ExtendedOperation) {
+    const span = getOtelTracer().startSpan(
+      "graphql-request",
+      {
+        kind: SpanKind.CLIENT,
+      },
+      context.active(),
+    );
+
+    span.setAttribute(
+      AttributeNames.OPERATION_NAME,
+      `${(operation.query.definitions[0] as Definition).name.value ?? "unknown"}`,
+    );
+
+    span.setAttribute(AttributeNames.OPERATION_TYPE, operation.kind);
+
+    span.setAttribute(AttributeNames.OPERATION_BODY, operation.query.loc?.source.body ?? "unknown");
+
+    span.setAttribute(AttributeNames.OPERATION_KEY, operation.key);
+
+    span.setAttribute(ObservabilityAttributes.SALEOR_API_URL, operation.context.url);
+
+    span.setAttribute(SemanticAttributes.HTTP_URL, operation.context.url);
+
+    addRequestHeaderAttributes(span, operation.context.fetchOptions?.headers);
+    if (operation.variables) {
+      addInputVariableAttributes(span, operation.variables);
+    }
+
+    return makeOperation(operation.kind, operation, {
+      ...operation.context,
+      span,
+    });
+  },
+  // @ts-ignore - small hack, we're extending `operation` with `span`
```

Comment:
**Suggestion:** maybe we can use `ts-expect-error` instead?

Docs: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-9.html#ts-ignore-or-ts-expect-error

We can also write just `@ts-expect-error we are extending operation with span`

Reply:
good idea, fixed